Add an image button component

Goldorion opened the pull request:
It adds the image button component everyone requests for too many years now.

The usage of a texture atlas is not supported, because I wasn't able to resolve all problems I had.

Note: GuiComponent has been re-ordered for more clarity when adding a new component.

KlemenDEV commented:
> The usage of a texture atlas is not supported, because I wasn't able to resolve all problems I had.

What does this mean in practice?

Goldorion commented:
> > The usage of a texture atlas is not supported, because I wasn't able to resolve all problems I had.
> 
> What does this mean in practice?

> The usage of a texture atlas is not supported

The image used never change even if you click on the button

> because I wasn't able to resolve all problems I had.

Due to the multiple possibilities of image selection a texture atlas allows, I could not find a solution to draw in the WYSIWYG editor the image chosen by the user (there is always one case where the image is invisible).

KlemenDEV commented:
You could crop the atlas to the default state of a non-pressed button for the WYSIWYG. Without pressed/focused images, this feature is not what users expect

Goldorion commented:
> You could crop the atlas to the default state of a non-pressed button for the WYSIWYG. Without pressed/focused images, this feature is not what users expect

This is already what I was trying to do, but I'm not able to get this image for all cases.

KlemenDEV commented:
> but I'm not able to get this image for all cases.

Not sure I understand this fully

Goldorion commented:
> > but I'm not able to get this image for all cases.
> 
> Not sure I understand this fully

In a texture atlas, you can use an image starting anywhere in the file and this is my problem. When the image start at a random point in the file, the image is invisible.

KlemenDEV commented:
What if you request multiple textures, like default, optional pressed, and optional focused and based on those, build an atlas image while storing the GE? 

This way, you know where the images are.

This feature was not added so far exactly due to a problem of figuring out how to make this exact thing work, so it is important we get it right if we decide to add this feature.

Goldorion commented:
> What if you request multiple textures, like default, optional pressed, and optional focused and based on those, build an atlas image while storing the GE?
> 
> This way, you know where the images are.
> 
> This feature was not added so far exactly due to a problem of figuring out how to make this exact thing work, so it is important we get it right if we decide to add this feature.

The second texture selector has been added. Thank you for the solution!

Goldorion commented:
I fixed the priority for both image buttons and checkboxes because I wasn't aware it was meant for their priority to when they should be drawn. I also added some doc for the method, so future contributors won't make the same mistake I did.

KlemenDEV commented on plugins/generator-1.19.2/forge-1.19.2/templates/gui/gui_window.java.ftl:
```diff
@@ -203,6 +203,28 @@ public class ${name}Screen extends AbstractContainerScreen<${name}Menu> {
+		<#list data.getComponentsOfType("ImageButton") as component>
+				this.addRenderableWidget(new ImageButton(this.leftPos + ${(component.x - mx/2)?int}, this.topPos + ${(component.y - my/2)?int},
+					${component.width}, ${component.height}, 0, 0, ${component.height}, new ResourceLocation("${modid}:textures/screens/atlas/${component.getName()}.png"), ${component.width}, ${component.height * 2}, e -> {
+							<#if hasProcedure(component.onClick)>
```
Code below this is same for button and image button, consider making eg. macro for this code part so it is not duplicated

KlemenDEV reviewed (changes requested):
Check the comment + please move the weight refactor to another PR so it can be reviewed and merged separately before the ImageButton PR. Thanks!

KlemenDEV reviewed (changes requested):
The hovered texture should be optional (if empty, default is used). Also apply this to tests, so both combinations are tested.

There should be a validator to make sure sizes match as it is quite easy to make one

Goldorion commented on src/main/java/net/mcreator/util/image/ImageUtils.java:
```diff
@@ -497,4 +497,26 @@ private static float limitDeviation(float original, float deviation, int sign) {
+	 * @param ySecond <p>The y position of the second image on the final image<</p>
+	 * @return <p>Returns the generated image.</p>
+	 */
+	public static BufferedImage merge2Images(Image first, Image second, int width, int height, int xFirst, int yFirst,
```
Oh... I thought 2 would mean 2 (two) 😅

Defeatomizer replied:
You can also go without the digit as image arguments are already named as ordinal numbers.

Goldorion commented:
> There should be a validator to make sure sizes match as it is quite easy to make one

I tried to add a converter, but it never worked (maybe because it uses 2 components), so I just added a check when making the texture atlas.

KlemenDEV commented:
There are merge conflicts to be fixed

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -571,6 +571,10 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
 				components.add(new Button("button3", 10, 10, "button3", 100, 200, null, new Procedure("condition3")));
 				components.add(new Button(AbstractWYSIWYGDialog.textToMachineName(components, null, "button"), 10, 10,
 						"button4", 100, 200, new Procedure("procedure2"), new Procedure("condition4")));
+				components.add(new ImageButton(0, 48, 16, 16, "picture1", "picture2", new Procedure("procedure10"), null));
```
Tests should only test cases where only one picture is specified

KlemenDEV reviewed (changes requested):
Changes still needed:

- merge with master (weights of components)
- Validation should be done in UI, not in code. SearchableComboBox is validable so this is doable
- buttons should have unique IDs as there may be future procedures addressing them

Goldorion commented:
> Changes still needed:
> 
>     * merge with master (weights of components)
> 
>     * Validation should be done in UI, not in code. SearchableComboBox is validable so this is doable
> 
>     * buttons should have unique IDs as there may be future procedures addressing them

I found out how to make the converter (and merged the master branch). However, concerning the unique button IDs, normal buttons don't have one, so why only these buttons would have an ID? (In addition to the fact, there are not a lot of useful methods and even fewer, maybe none, for the average MCreator user (the useful ones are quite specific))

KlemenDEV commented on src/main/java/net/mcreator/element/parts/gui/GUIComponent.java:
```diff
@@ -77,6 +78,11 @@
+	/**
+	 * Returns the priority for when this component should be drawn in the UI, to represent how Minecraft draws components in the game.
+	 *
+	 * @return The priority of the component (lower is more important, higher is less important)
```
> lower is more important, higher is less important

This is not correct. Lower means it will be rendered closer to the screen and higher means it will "sink" more behind other components

KlemenDEV commented:
>  However, concerning the unique button IDs, normal buttons don't have one

This is unique name:

![image](https://user-images.githubusercontent.com/16374228/216692783-1abeaf89-d681-4bf1-bcdd-5ebbe567adcb.png)

> In addition to the fact, there are not a lot of useful methods and even fewer, maybe none, for the average MCreator user (the useful ones are quite specific)

For this reason:

> as there may be future procedures addressing them

Eg: change image, disable, ...

It is easier to make it right now than to make converter to "uniquefy" names later.

It literally takes a few lines of changes to do it right now

KlemenDEV commented on src/main/java/net/mcreator/element/parts/gui/ImageButton.java:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	public Image getHoveredImage(Workspace workspace) {
+		if (hoveredImage != null && !hoveredImage.isEmpty()) {
```
Given you call getImage 3 times and always at leas one time, this should be a variable so getImage logic is not executed 3 times

KlemenDEV commented on src/main/java/net/mcreator/ui/validation/validators/ImageSizeMatchValidator.java:
```diff
@@ -0,0 +1,87 @@
+	 * This validator takes 2 images from a same {@link TextureType} and checks if they have the same width and height.
+	 *
+	 * @param firstImage        This is the first image of the check. This image can never be null or empty.
+	 * @param secondImage This
```
You are missing text here as it only writes This

Goldorion replied:
Oops 😬

KlemenDEV commented on src/main/java/net/mcreator/ui/validation/validators/ImageSizeMatchValidator.java:
```diff
@@ -0,0 +1,87 @@
+	 * @param workspace
+	 * @param passIfSecondEmpty
+	 */
+	public ImageSizeMatchValidator(String firstImage, String secondImage, TextureType type, Workspace workspace,
```
The validator should take components, not values. setFirstImage and setSecondImage should not be used, instead, components should be passed. 

Check other validators implemented for inspiration. As this validator will very unlikely be used in other components, it could also be inlined in anonymous class

Goldorion commented:
> as there may be future procedures addressing them

So normal buttons also need to be changed as they extends the Button class, the class used by normal buttons.

KlemenDEV commented:
> So normal buttons also need to be changed as they extends the Button class, the class used by normal buttons.

No, why? ID is not for the code, but internal reference for MCreator that we may use later, as explained in my comment above. I have shown you where ID comes from in the screenshot above. It is just so every image button has unique name/ID in MCreator

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/wysiwyg/ImageButtonDialog.java:
```diff
@@ -0,0 +1,170 @@
+			ImageIcon image2 = editor.mcreator.getWorkspace().getFolderManager()
+					.getTextureImageIcon(FileUtils.removeExtension(secondTexture), TextureType.SCREEN);
+
+			if (image1.getIconHeight() == image2.getIconHeight() && image1.getIconWidth() == image2.getIconWidth())
```
Didn't you add a method to check this in ImageUtils?
